Built-in Actor events for DDO

aarshkshah1992 opened the pull request:
Closes https://github.com/filecoin-project/builtin-actors/issues/1464.

This PR implements [FIP-0083](https://github.com/filecoin-project/FIPs/blob/7eb8da9f5d50f2664e059bd56c930272d6189a57/FIPS/fip-0083.md). The corresponding FIP discussion is [here](https://github.com/filecoin-project/FIPs/discussions/754).

This PR implements built-in Actor events specified in FIP-0083 on top of the [DDO branch](https://github.com/filecoin-project/builtin-actors/pull/1486).

### TODO
- [x] Get all integration tests green
- [x] Gas costs estimation for event emission
- [x] Merge https://github.com/filecoin-project/FIPs/pull/907
- [x] Wait for #1486 to land on master before merging.
- [x] Merge FIP change https://github.com/filecoin-project/FIPs/pull/913
- [x] Merge FIP change https://github.com/filecoin-project/FIPs/pull/923
- [x] Wait for https://github.com/filecoin-project/builtin-actors/pull/1483 to land into master

aarshkshah1992 commented on actors/miner/src/lib.rs:
```diff
@@ -4795,6 +4848,31 @@ fn verify_deals(
+    rt: &impl Runtime,
+    deal_id: &DealID,
+) -> Result<ext::market::GetDealDataCommitmentReturn, ActorError> {
+    deserialize_block(extract_send_result(rt.send_simple(
```
@anorth 

I think we should add a batched version of this method to the Market Actor so that the Miner Actor can lookup piece information for a array of deal_ids. This will reduce the number of calls we make to the Market Actor for emitting Sector Activation/Update events. Will have to change the FIP as well to add this new `GET_DEAL_DATA_COMMITMENT_BATCHED` method to the Market Actor ?

aarshkshah1992 commented on actors/miner/src/lib.rs:
```diff
@@ -4795,6 +4848,31 @@ fn verify_deals(
+) -> Result<ext::market::GetDealDataCommitmentReturn, ActorError> {
+    deserialize_block(extract_send_result(rt.send_simple(
+        &STORAGE_MARKET_ACTOR_ADDR,
+        ext::market::GET_DEAL_DATA_COMMITMENT,
```
@anorth 

I think we should add a batched version of this method to the Market Actor so that the Miner Actor can lookup piece information for a array of deal_ids. This will reduce the number of calls we make to the Market Actor for emitting Sector Activation/Update events. Will have to change the FIP as well to add this new `GET_DEAL_DATA_COMMITMENT_BATCHED` method to the Market Actor ?

aarshkshah1992 replied:
@anorth I have implemented Option #2 mentioned in https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-8071837. I've updated this PR with the change. Please can you take a look ?

anorth commented on test_vm/src/messaging.rs:
```diff
@@ -637,8 +642,11 @@ impl<'invocation> Runtime for InvocationCtx<'invocation> {
     }
 
     // TODO No support for events yet.
```
Stale comment

anorth commented on vm_api/src/trace.rs:
```diff
@@ -2,10 +2,17 @@ use fvm_ipld_encoding::ipld_block::IpldBlock;
+pub struct EmittedEvent {
+    pub emitter: ActorID,
+    pub event: ActorEvent,
+}
```
These changes will need to be propagated to https://github.com/anorth/fvm-workbench/blob/main/api/src/trace.rs, pending us moving this API to somewhere it can be imported by all (cc @Stebalien ). You'll probably immediately run into this need when doing gas tracing.

aarshkshah1992 replied:
Will do this as part of the FVM workbench gas analysis.

Stebalien replied:
Yeah, storing events in the trace seems reasonable.

anorth reviewed (changes requested):
Most of this looks great, but calling out to the market actor to compute event fields is a blocker.

rvagg commented on actors/miner/src/emit.rs:
```diff
@@ -0,0 +1,79 @@
+pub fn sector_activated(
+    rt: &impl Runtime,
+    sector: SectorNumber,
+    unsealed_cid: &Cid,
```
Seems reasonable since they're already an `Option` and we force the receiver of the event to special-case these zero-cids too. Are the zero-cids exposed publicly yet at all or are they just an internal detail? Why do we need them like this internally anyway? Are there special cases that require it to be present?

anorth replied:
They're just precomputed for efficiency. They are public and meaningful as signifying a sector empty of data. There are processes that check for equality with the zero CID

aarshkshah1992 commented on integration_tests/src/expects.rs:
```diff
@@ -343,4 +366,76 @@ impl Expect {
+
+    pub fn build_verifreg_event(
+        typ: &str,
+        id: &u64,
```
Yeah because copies of primitive types are cheap. Makes sense !

aarshkshah1992 commented on actors/market/tests/harness.rs:
```diff
@@ -387,10 +413,23 @@ pub fn batch_activate_deals_raw(
     provider: Address,
     sectors_deals: Vec<SectorDeals>,
     compute_cid: bool,
+    expected_activated_deals: Vec<DealID>,
```
@anorth I tried hacking this but it's tricky  because the current mockruntime event emission code immediately matches the event being emitted against the latest expected event. See here:

https://github.com/filecoin-project/builtin-actors/blob/d06215ba1e7391a2053684f5bcee20d935711cc1/runtime/src/test_utils.rs#L1261

I tried refactoring this so that all expected events and emitted events are checked together in the end instead being checked during emission but it needs some more refactor in the mock runtime. I think we can punt this for now.

anorth replied:
Ok, can you please write an issue for that, summarising both our points above, and then link to that from here?

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -5513,15 +5581,28 @@ fn activate_sectors_deals(
+            for unverified_info in &sector_deals.unverified_infos {
+                piece_infos_for_sector.push((unverified_info.data, unverified_info.size.0));
+            }
+            piece_infos.push(piece_infos_for_sector);
```
This roundtrip to the market actor and returning the pieces in two lists, verified and unverified, means we've lost the ordering of pieces within the sector. It's not obvious how much of a problem that might, be but we should be able to retain it. We'll need to modify the market actor return value to maintain the order of the deals.

anorth commented on actors/miner/src/lib.rs:
```diff
@@ -5427,18 +5491,20 @@ fn activate_sectors_pieces(
     activation_infos: &[DealsActivationInput],
     compute_unsealed_cid: bool,
-) -> Result<(BatchReturn, Vec<DataActivationOutput>), ActorError> {
+) -> Result<(BatchReturn, Vec<DataActivationOutput>, Vec<Vec<(Cid, u64)>>), ActorError> {
```
The point of this DataActivationOutput type is to carry whatever per-sector activation metadata is needed by other processes. We should include the piece information in there rather than need to accumulate it in a separate map and look it up again later. That will remove some hard-to-mentally-verify indexing and unwraps.

anorth reviewed (commented):
The market actor change came out nice and simple. I think we can clean up the data flow a little though. I will attempt to do this for you today.

anorth commented:
I've pushed a commit implementing my own suggestions. Please check it carefully.

I'm approving, but please don't merge this to the integration/direct-onboarding branch. We'll rebase it on master after the former lands.

aarshkshah1992 commented:
@anorth Your changes are great and lgtm. FIP update raised at https://github.com/filecoin-project/builtin-actors/pull/1491.

Stebalien commented on actors/market/src/lib.rs:
```diff
@@ -482,6 +483,13 @@ impl Actor {
+
+            emit::deal_published(
+                rt,
+                valid_deal.proposal.client.id().unwrap(),
+                valid_deal.proposal.provider.id().unwrap(),
```
Are you sure we guarantee that these are IDs? Especially the client?

I see here that we resolve the client:

https://github.com/filecoin-project/builtin-actors/blob/f8916da127502cb2dd24651466cd89a2ffbb1f38/actors/market/src/lib.rs#L291-L300

aarshkshah1992 replied:
@Stebalien Yeah this is fine because we do not add deal proposals with non-id clients to the `valid_deals` slice. See [here](https://github.com/filecoin-project/builtin-actors/blob/7e0abcacc4ad8c1d1ff4ef404f9f173380388acf/actors/market/src/lib.rs#L296) AND [here](https://github.com/filecoin-project/builtin-actors/blob/7e0abcacc4ad8c1d1ff4ef404f9f173380388acf/actors/market/src/lib.rs#L383).

The `valid_deals` slice is then used to emit these events.

Stebalien replied:
Ah, the relevent code is https://github.com/filecoin-project/builtin-actors/blob/7e0abcacc4ad8c1d1ff4ef404f9f173380388acf/actors/market/src/lib.rs#L336-L337

Stebalien commented on actors/market/src/state.rs:
```diff
@@ -740,7 +740,7 @@ impl State {
         deal: &DealProposal,
         epoch: ChainEpoch,
-    ) -> Result<(TokenAmount, bool), ActorError>
+    ) -> Result<(TokenAmount, bool, bool), ActorError>
```
I'm still confused. Is there documentation somewhere?

Stebalien replied:
nvm. I'm not sure why the documentation wasn't showing up...

Stebalien commented on actors/market/src/lib.rs:
```diff
@@ -482,6 +483,13 @@ impl Actor {
+                rt,
+                valid_deal.proposal.client.id().unwrap(),
+                valid_deal.proposal.provider.id().unwrap(),
+                new_deal_ids[i],
```
nit: put `let deal_id = new_deal_ids[i]` at the top of the loop (it's used twice). Even better, change the loop to

```rust
for (valid_deal, &deal_id) in valid_deals.iter().zip(&new_deal_ids) {
```

Stebalien commented on actors/market/src/lib.rs:
```diff
@@ -54,6 +54,7 @@ pub mod policy;
 pub mod testing;
 
 mod deal;
+pub mod emit;
```
Is this supposed to be public?

Stebalien replied:
nit: I'd also call it `events`, or `event_utils`, or something like that.

Stebalien replied:
But not critical if that causes an annoying rewrite.

Stebalien replied:
Eh, actually, I see why you're calling this emit. Nvm.

Stebalien commented on actors/market/src/lib.rs:
```diff
@@ -635,11 +637,17 @@ impl Actor {
-                });
+                activations.push(SectorDealActivation { activated, unsealed_cid: data_commitment });
+
+                for (deal_id, proposal) in sector.deal_ids.iter().zip(&validated_proposals) {
```
Any reason not to combine this with the previous for loop?

aarshkshah1992 replied:
Just makes it easy to see that events are only emitted after everything else has been handled. For example, see [here](https://github.com/filecoin-project/builtin-actors/blob/7e0abcacc4ad8c1d1ff4ef404f9f173380388acf/actors/market/src/lib.rs#L631).

Stebalien replied:
Sg.

Stebalien commented on actors/miner/src/lib.rs:
```diff
@@ -1006,24 +1026,25 @@ impl Actor {
             .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load sectors array")?;
         let mut sector_infos = Vec::with_capacity(params.sector_updates.len());
         let mut updates = Vec::with_capacity(params.sector_updates.len());
+        let mut sector_commds: HashMap<SectorNumber, CompactCommD> = HashMap::new();
```
nit: `with_capacity(params.sector_updates.len()`

aarshkshah1992 commented on actors/miner/src/lib.rs:
```diff
@@ -90,6 +91,8 @@ pub mod testing;
 mod types;
 mod vesting_state;
 
+pub mod emit;
```
No it does not. You are right.

aarshkshah1992 commented on actors/miner/src/lib.rs:
```diff
@@ -1972,11 +2012,17 @@ impl Actor {
-        )
+        )?;
+
+        for (pc, data) in successful_activations.iter().zip(data_activations.iter()) {
+            let unsealed_cid = pc.info.unsealed_cid.0;
+            emit::sector_activated(rt, pc.info.sector_number, unsealed_cid, &data.pieces)?;
+        }
```
`activate_new_sector_infos ` is also used by some other methods that have slightly different data flows. So, best to keep it here.

rvagg commented on actors/market/src/lib.rs:
```diff
@@ -554,7 +562,7 @@ impl Actor {
             let mut batch_gen = BatchReturnGen::new(params.sectors.len());
             let mut activations: Vec<SectorDealActivation> = vec![];
-            let mut activated_deals = BTreeSet::<DealID>::new();
+            let mut activated_deals: HashSet<DealID> = HashSet::new();
```
```suggestion
            let mut activated_deals = HashSet::<DealID>::new();
```

rvagg replied:
or not .. I see some inconsistency below, like in actors/miner/tests/util.rs where there's a lot of explicit variable typing so I guess we don't have a rule in here (yet)

rvagg commented on actors/miner/tests/util.rs:
```diff
@@ -1304,6 +1347,23 @@ impl ActorHarness {
+            {
+                continue;
+            }
+            let pieces: Vec<(Cid, u64)> = sm.pieces.iter().map(|p| (p.cid, p.size.0)).collect();
```
```suggestion
            let pieces = sm.pieces.iter().map(|p| (p.cid, p.size.0)).collect();
```

rvagg commented on actors/miner/tests/util.rs:
```diff
@@ -1488,6 +1552,23 @@ impl ActorHarness {
+            {
+                continue;
+            }
+            let pieces: Vec<(Cid, u64)> = sm.pieces.iter().map(|p| (p.cid, p.size.0)).collect();
```
```suggestion
            let pieces = sm.pieces.iter().map(|p| (p.cid, p.size.0)).collect();

```

anorth commented on actors/market/src/lib.rs:
```diff
@@ -926,13 +942,14 @@ impl Actor {
+                            curr_epoch,
+                        )?;
 
                     if remove_deal {
```
I think this should emit deal-terminated, immediately after the `provider_deals_to_remove` update at 963. This will mimic the same code in the synchronous case. Yes, this is inside cron, but the FIP says we will emit the event here.

aarshkshah1992 replied:
Added the corresponding here but note that it will only be emitted if the deal is not already completed as we don't want to emit both "completed" and "terminated" events in this case.

anorth commented on actors/miner/src/ext.rs:
```diff
@@ -40,16 +42,16 @@ pub mod market {
 
     #[derive(Serialize_tuple, Deserialize_tuple, Clone)]
-    pub struct VerifiedDealInfo {
+    pub struct ActivateDealInfo {
```
Please match the original name of `ActivatedDeal`.

aarshkshah1992 replied:
Done.

anorth commented on actors/miner/src/ext.rs:
```diff
@@ -40,16 +42,16 @@ pub mod market {
-    impl Default for VerifiedDealInfo {
-        fn default() -> VerifiedDealInfo {
-            VerifiedDealInfo {
+    impl Default for ActivateDealInfo {
```
Nit: do we need default? Can we nix it?

aarshkshah1992 replied:
Yes removed it.

anorth commented on integration_tests/src/expects.rs:
```diff
@@ -42,6 +45,8 @@ impl Expect {
     pub fn market_activate_deals(
         from: ActorID,
         deals: Vec<DealID>,
+        // TODO: A more future-proof one would take the deals as a Vec<(DealID, ActorId)>
```
Nit: remove the word "TODO". No-one's going to do it until they're forced to. Leaving a note here is great, but "TODO" doesn't add any information.

aarshkshah1992 replied:
Removed this.

anorth reviewed (changes requested):
I'm afraid the rebase against #1313 was too hasty and has missed some critical changes. That PR added a new method for deal settlement to the market, and that method must emit events for deal completion and termination. 

Also I note the commit history for this PR is long and complicated, including commit corresponding to many already landed on master. These might disappear with the appropriate rebase incantations. If you can, please squash down so that there is just one commit per previously-reviewed PR that was merged into the integration branch. We can then land on master by rebasing. If that proves difficult, we'll land by squashing so you may as well squash everything to a single commit now. (Why is this PR not a merge of `integration/actor-events` into master?)

aarshkshah1992 commented on actors/market/src/lib.rs:
```diff
@@ -1247,7 +1273,7 @@ impl Actor {
                 }
 
-                let (_, payment_amount, remove_deal) = match st.process_deal_update(
+                let (_, payment_amount, _, remove_deal) = match st.process_deal_update(
```
Apologies, missed this. Added both deal completion and deal termination events here.

aarshkshah1992 commented:
@anorth Addressed your review and CI is green. Let me know what you think.

This PR is not a merge of `integration/actor-events` into master as that branch was originally created for events in a non-DDO world. So, I moved to this new branch once we decided to re-do the events work against the DDO branch.

anorth commented on actors/miner/src/ext.rs:
```diff
@@ -40,29 +42,16 @@ pub mod market {
 
     #[derive(Serialize_tuple, Deserialize_tuple, Clone)]
-    pub struct VerifiedDealInfo {
+    pub struct ActivateDeal {
```
Activate**d**Deal
